x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -394,10 +396,14 @@ protected void asyncShardOperation(
         protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
             ActionListener<Response> wrappedListener = ActionListener.wrap(listener::onResponse, e -> {
                 Throwable cause = ExceptionsHelper.unwrapCause(e);
-                if (cause instanceof IllegalStateException && cause.getMessage().contains("Not all operations between from_seqno [")) {
+                if (cause instanceof MissingHistoryOperationsException) {
```

Comment:
I don't think this is serializing properly. I find this exception at a breakpoint here. I think that our exception serializing code is doing an `instanceof` check, seeing that this is an `IllegalStateException` and it is identified as an `IllegalStateException` on deserialization.

```
java.lang.IllegalStateException: Not all operations between from_seqno [17] and to_seqno [34] found; expected seqno [17]; found [Index{id='6', type='doc', seqNo=18, primaryTerm=1, version=2, autoGeneratedIdTimestamp=-1}]
	at org.elasticsearch.index.engine.LuceneChangesSnapshot.rangeCheck(LuceneChangesSnapshot.java:155)
	at org.elasticsearch.index.engine.LuceneChangesSnapshot.next(LuceneChangesSnapshot.java:138)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction.getOperations(ShardChangesAction.java:527)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction$TransportAction.shardOperation(ShardChangesAction.java:340)
	at org.elasticsearch.xpack.ccr.action.ShardChangesAction$TransportAction.shardOperation(ShardChangesAction.java:319)
	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$1.doRun(TransportSingleShardAction.java:117)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:732)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
```

Reply:
I made a wrong assumption here. I thought that the exception here didn't get serialized, but it does get serialized. (when the request gets send to a different shard copy then the local node has).

This means that throwing the `MissingHistoryOperationsException` here does not work. We need to throw an exception that extends from `ElasticticsearchException` and that sets specific metadata in the LuceneSnapshot.java . I will make this change now.